elasticapm/base.py

```diff
@@ -360,12 +360,42 @@ def get_system_info(self):
             system_data["kubernetes"] = k8s
         return system_data
 
+    def get_cloud_info(self):
+        """
+        Detects if the app is running in a cloud provider and fetches relevant
+        metadata from the cloud provider's metadata endpoint.
+        """
+        provider = self.config.cloud_provider
+
+        if provider is False:
+            return {}
+        if provider == "aws":
+            return cloud.aws_metadata()
+        elif provider == "gcp":
+            return cloud.gcp_metadata()
+        elif provider == "azure":
+            return cloud.azure_metadata()
+        else:
+            # Trial and error
+            data = {}
+            data = cloud.aws_metadata()
+            if data:
+                return data
+            data = cloud.gcp_metadata()
+            if data:
+                return data
+            data = cloud.azure_metadata()
+            return data
+
     def build_metadata(self):
         data = {
             "service": self.get_service_info(),
             "process": self.get_process_info(),
             "system": self.get_system_info(),
+            "cloud": self.get_cloud_info(),
```

Comment:
++ on background thread. DNS not having timeouts is a bummer :(

Reply:
I actually don't think any of these concerns are an issue -- all of the metadata servers are host local, there's no actual wire activity involved. Additionally, none of the addresses are actually routable so they fail immediately if the host is not in the given cloud provider.

That said, it might be trivial to move metadata generation into the beginning of the transport thread. Since we're generating metadata right at thread start anyway, it could just be just after thread start, as opposed to just before. I'll try this.

Reply:
Done.